### ISSUE TITLE
Overloading on borrowing/consuming results in ambiguities with performance hacks removed  

### ISSUE THREAD

**Issue by xedin** [swiftlang/swift#77001]
### Description

[One of the test-cases](https://github.com/swiftlang/swift/pull/76834/files#diff-5da86c6657a1dc4926cc9b107f6c45cf4fcd86c54cb102d31c7821feda90eb17R65) introduced by https://github.com/swiftlang/swift/pull/76834 depends on a type-checker hack that applies only to initializers - https://github.com/swiftlang/swift/blob/main/lib/Sema/CSSimplify.cpp#L9843-L9855. Removal of this hack makes the code ambiguous.

### Reproduction

```swift
struct View : ~Escapable {
 init(_ otherBV: borrowing View) {}
 init(_ k: consuming View) {}
}

@lifetime(x)
func getConsumingView(_ x: consuming View) -> View {
  return View(x)
}
```

Without perf hack this produces:

```
 error: ambiguous use of 'init(_:)'
```

### Expected behavior

We need to define expected type/flag matching behavior for `borrowing` and `consuming` parameters that does not depend on the performance hack(s). 

### Environment

Swift compiler main branch.

**Comment by xedin** [comment]
cc @meg-gupta

**Comment by xedin** [comment]
This is actually even worse than just ambiguity because the hack always prefers the first overload which means that flipping overloads around would result in a different pick.

**Comment by meg-gupta** [comment]
This is an issue with non initializers as well. 

This is accepted: 
```
class Klass {}

func test(_ a: borrowing Klass) {}
func test(_ a: consuming Klass) {}
```

**Comment by jckarter** [comment]
These should be rejected. We don't allow overloading on ownership qualifiers.

**Comment by xedin** [comment]
@meg-gupta Yeah, similar hack for one parameter overload sets, but that falls apart when there is more than one.
